Implement a way to time out tokens based on (in)activity

simo5 opened the pull request:
When OAuthClient is configure with accessTokenTimeoutSeconds then tokens obtained with the specific client get a TimeoutsIn field that marks when the token is to be considered timed out.
The timeout is in seconds since token CreationTimestamp
As the token is used it is pushed into a bucket which is regularly flushed (for now).

This replaces #17161 which github inexplicably and unilaterally decided to close after a wrong push (can't be reopened)

simo5 commented:
/hold
need to wait for api merge in openshift/api

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+	delta := a.clientTimeout(td.token.ClientName)
+	newTimeout := int32(0)
+	if delta > 0 {
+		newTimeout = int32((td.seen.Sub(td.token.CreationTimestamp.Time) + delta) / time.Second)
```
I had some comments explaining this math which I think are useful.

simo5 replied:
added

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+	if oauthClient.AccessTokenTimeoutSeconds == nil {
+		return a.defaultTimeout
+	}
+	a.updateTimeouts(oauthClient)
```
I believe I had some extra glogs in this func and some other places in this file?

simo5 replied:
I did not import all of them as some things simply changed, any specific thing you want to generate logs out of ? You can describe the thought process about what you think should be logged too.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+		switch {
+		case err == nil:
+			flushedTokens++
+		case apierrors.IsConflict(err):
```
Combine the two error cases.

simo5 replied:
done

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
+				td.token.UserName, td.token.ClientName, td.token.Scopes, err)
+		} else {
+			flushedTokens++
```
Should we also count total failures?

simo5 replied:
Any token not flushed is a failure, do we need to be very explicit about it ?

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+		}
+	}
+
+	glog.Infof("Flushed %d tokens out of %d in bucket", flushedTokens, totalTokens)
```
ack

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+	glog.Infof("Flushed %d tokens out of %d in bucket", flushedTokens, totalTokens)
+}
+
+func (a *TimeoutValidator) Start(stopCh <-chan struct{}) {
```
I'll take this as "we need unit testing", I am not in love with clock injection, will think more about this point

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+	// make sure to kill the timer when we exit
+	defer closeTimer()
+
+	updateTimerAndFlush := func() {
```
good point, done

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,267 @@
+			tokenTimeout := td.timeout()
+			safetyTimeout := nextTimeout.Add(a.safetyMargin)
+			if safetyTimeout.After(tokenTimeout) {
+				glog.Infof("Timeout for user=%q client=%q scopes=%v falls below safety margin (%s < %s) forcing flush",
```
ack

simo5 commented on pkg/cmd/server/origin/authenticator.go:
```diff
@@ -92,6 +94,14 @@ func newAuthenticator(config configapi.MasterConfig, accessTokenGetter oauthclie
+			timeoutValidator := authnregistry.NewTimeoutValidator(accessTokenGetter, oauthClientLister, *config.OAuthConfig.TokenConfig.AccessTokenTimeoutSeconds)
+			validators = append(validators, timeoutValidator)
+			postStartHooks["openshift.io-TokenTimeoutUpdater"] = func(context genericapiserver.PostStartHookContext) error {
+				go timeoutValidator.Start(context.StopCh)
```
Ok, changed Start() -> Run() and added a HandleCrash call in it

simo5 commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -17,6 +17,7 @@ import (
 )
 
 const MinTokenLength = 32
+const MinFlushTimeout = 150
```
ack

simo5 replied:
I am actually going to change flush timeout handling a little, and will put here docs on what it means

simo5 commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -81,6 +82,11 @@ func ValidateAccessTokenUpdate(newToken, oldToken *oauthapi.OAuthAccessToken) fi
 	allErrs := validation.ValidateObjectMetaUpdate(&newToken.ObjectMeta, &oldToken.ObjectMeta, field.NewPath("metadata"))
 	copied := *oldToken
 	copied.ObjectMeta = newToken.ObjectMeta
+	// allow only TimeoutsIn to be changed
```
Ok I think I have it now

simo5 commented on test/integration/oauth_timeout_test.go:
```diff
@@ -0,0 +1,200 @@
+			AccessTokenTimeoutSeconds: &invalid,
+			GrantMethod:               oauthapi.GrantHandlerAuto,
+		})
+		if err == nil {
```
done

simo5 commented:
Ok I should have addressed all comments but the clock one.
I changed slightly how the defaultTimeout handling works.
Next a change to how we change the flushTimeout, current code can only make it shrink but never grow back. going to change this so we can recover after an admin fat fingered a too short interval on an oauthClient configuration

simo5 commented:
Finally #17662 got in and I was able to properly generate the internal API changes, code has been updated

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,298 @@
+	"github.com/openshift/origin/pkg/user/apis/user"
+)
+
+var ErrTimedout = errors.New("token timed out")
```
nit: I believe this can be made private.

simo5 replied:
ok

enj commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -102,6 +103,16 @@ func ValidateOAuthConfig(config *api.OAuthConfig, fldPath *field.Path) Validatio
 	}
 
+	if config.TokenConfig.AccessTokenTimeoutSeconds != nil {
+		if *config.TokenConfig.AccessTokenTimeoutSeconds < oauthvalidation.ReferenceTimeoutInterval {
```
*config.TokenConfig.AccessTokenTimeoutSeconds == 0 is valid

simo5 replied:
fixed

enj commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -79,8 +89,24 @@ func ValidateAccessToken(accessToken *oauthapi.OAuthAccessToken) field.ErrorList
+	// to timeout earlier than they should. 0 is the exception, as this value
+	// indicates that the token does not expire, and it is an allowed new
+	// value.
+	if newToken.TimeoutsIn > 0 && newToken.TimeoutsIn < oldToken.TimeoutsIn {
```
I had some related validation in `pkg/oauth/registry/oauthaccesstoken/strategy.go` that went with this as well.

simo5 replied:
Yeah I saw that, but given we can update the field with any (valid) value, I do not see much value in adding this kind of validation just at token creation.

enj replied:
OK I suppose that is fine.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -16,30 +15,27 @@ import (
 }
 
-func (a *tokenData) timeout() time.Time {
+func (a tokenData) timeout() time.Time {
```
You probably want to use *tokenData to avoid copying the struct unnecessarily.

simo5 replied:
I was doing that before, but when I added RankedSet the compiler complained, changing it to not be apointer made it work.

enj replied:
Right because you will have to update everything to use `*tokenData` otherwise it will not implement the interface.

enj commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -102,6 +103,17 @@ func ValidateOAuthConfig(config *api.OAuthConfig, fldPath *field.Path) Validatio
 			)))
 	}
 
+	if config.TokenConfig.AccessTokenTimeoutSeconds != nil {
```
`if timeout := config.TokenConfig.AccessTokenTimeoutSeconds; timeout != nil && *timeout != 0 {` is a bit easier to read.

simo5 replied:
ok

enj commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -79,8 +89,24 @@ func ValidateAccessToken(accessToken *oauthapi.OAuthAccessToken) field.ErrorList
+			fmt.Sprintf("cannot beless than the current value=%d", oldToken.TimeoutsIn)))
+	}
+	// negative values are not allowed either
+	if newToken.TimeoutsIn < 0 {
```
We want this on create as well.

simo5 replied:
indeed

enj commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -139,6 +165,15 @@ func ValidateClient(client *oauthapi.OAuthClient) field.ErrorList {
 		allErrs = append(allErrs, ValidateScopeRestriction(restriction, field.NewPath("scopeRestrictions").Index(i))...)
 	}
 
+	if client.AccessTokenTimeoutSeconds != nil {
```
Want negative check here as well.

enj replied:
And same checks on update too I believe.

simo5 replied:
yeah about negative, update automatically checks for anything doen at create

enj commented:
@simo5 please add unit tests for all the validation bits, should be easy to extend/copy existing tests.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+		case err == nil:
+			flushedTokens++
+		case apierrors.IsConflict(err) || apierrors.IsServerTimeout(err):
+			glog.V(5).Infof("Token update deferred for token=%q, retriable error: %v",
```
here

enj replied:
Err actually I see you are logging the token name, do not do that xD

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+				td.token.Name, err)
+			retryList = append(retryList, *td)
+		default:
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
```
here

simo5 replied:
no token name here

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+	for _, td := range retryList {
+		err := a.update(td)
+		if err != nil {
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
```
here

simo5 replied:
neither here, is the err what you want me drop ? (doing that anyway)

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+		}
+	}
+
+	glog.V(5).Infof("Successfully flushed %d tokens out of %d",
```
You may want a.data.Len() as well?

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+		err := a.update(*td)
+		switch {
+		case err == nil:
+			flushedTokens++
```
A glog here may be useful.

simo5 replied:
why ?

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+			glog.V(5).Infof("Token timeout for user=%q client=%q scopes=%v was not updated: %v",
+				td.token.UserName, td.token.ClientName, td.token.Scopes, err)
+		} else {
+			flushedTokens++
```
A glog here may be useful.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+		return a.defaultTimeout
+	}
+	if oauthClient.AccessTokenInactivityTimeoutSeconds == nil {
+		return a.defaultTimeout
```
A glog here may be useful.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+	if oauthClient.AccessTokenInactivityTimeoutSeconds == nil {
+		return a.defaultTimeout
+	}
+	return timeoutAsDuration(*oauthClient.AccessTokenInactivityTimeoutSeconds)
```
A glog here may be useful.

enj commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -79,8 +94,24 @@ func ValidateAccessToken(accessToken *oauthapi.OAuthAccessToken) field.ErrorList
 
 func ValidateAccessTokenUpdate(newToken, oldToken *oauthapi.OAuthAccessToken) field.ErrorList {
 	allErrs := validation.ValidateObjectMetaUpdate(&newToken.ObjectMeta, &oldToken.ObjectMeta, field.NewPath("metadata"))
+	// since TimeoutsIn can be concurrently updated by multipe masters we do
```
We should error if `oldToken.TimeoutsIn == 0 && newToken.TimeoutsIn != 0` to prevent a token that was never supposed to timeout from being updated to timeout.  Probably needs a comment describing how this race could happen between multiple masters.

simo5 replied:
yes

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,278 @@
+
+var errTimedout = errors.New("token timed out")
+
+// Implements rankedset.Item
```
done

simo5 commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -102,6 +103,16 @@ func ValidateOAuthConfig(config *api.OAuthConfig, fldPath *field.Path) Validatio
+		if *timeout != 0 && *timeout < oauthvalidation.ReferenceTimeoutInterval {
+			validationResults.AddErrors(field.Invalid(
+				fldPath.Child("tokenConfig", "accessTokenTimeoutSeconds"),
+				*config.TokenConfig.AccessTokenInactivityTimeoutSeconds,
```
ok

simo5 commented on test/integration/oauth_timeout_test.go:
```diff
@@ -0,0 +1,200 @@
+	// Use the server and CA info
+	anonConfig := restclient.AnonymousClientConfig(clientConfig)
+
+	min := int32(150)
```
it passes tests ...

simo5 replied:
hint the min timeout is reference timeout / 3 ie 100 seconds is the current absolute minimum

simo5 commented:
/retest

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -228,33 +175,21 @@ func (a *TimeoutValidator) flush(flushHorizon time.Time) {
+func (a *TimeoutValidator) nextTick() time.Time {
+	// Add a small safety Margin so flushes tend to
+	// overlap a little rather than have gaps
+	return time.Now().Add(a.tickerInterval + time.Duration(10)*time.Second)
```
I believe `return time.Now().Add(a.tickerInterval + 10*time.Second)` work here so you do not need the type cast.

enj commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -265,15 +200,15 @@ func (a *TimeoutValidator) Run(stopCh <-chan struct{}) {
-		case <-nextTimer.C:
-			updateTimerAndFlush()
+		case <-ticker.C:
+			nextTick := a.nextTick()
```
`nextTick = a.nextTick()`, no `:=`

![image](https://user-images.githubusercontent.com/7001223/34016366-965ff9ce-e0ef-11e7-9b14-4d02287d6bc0.png)

simo5 commented:
Blocked until https://github.com/openshift/api/pull/15 is merged.

simo5 commented:
/retest

enj commented on pkg/auth/oauth/registry/registry_test.go:
```diff
@@ -473,3 +476,204 @@ func TestAuthenticateTokenValidated(t *testing.T) {
+	}
+}
+
+func TestAuthenticateTokenTimeout(t *testing.T) {
```
I think we will want a test for retry logic, and another test that "counts" reads/writes to ETCD (will probably have to be fuzzy).

simo5 replied:
add followup issue if you think we really need it, I am not sure we do

enj commented:
For anyone interested in @openshift/sig-security, feel free to take a look.

/assign @liggitt 

Jordan, it is up to you how much you want to dig into the implementation of this.  I can certainly explain it in person if you just prefer to just know the design details.

simo5 commented:
Now waiting on #17810

simo5 commented on pkg/auth/oauth/registry/registry_test.go:
```diff
@@ -473,3 +476,204 @@ func TestAuthenticateTokenValidated(t *testing.T) {
+	return list, nil
+}
+
+func checkToken(t *testing.T, name string, authf authenticator.Token, tokens oauthclient.OAuthAccessTokenInterface, present bool) {
```
I see no use of t.Helper() except for one place in vendor/
Is it about just calling it once at the start of the function?

simo5 commented on pkg/auth/oauth/registry/registry_test.go:
```diff
@@ -473,3 +476,204 @@ func TestAuthenticateTokenValidated(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	defaultTimeout := int32(30) // 30 seconds
```
Yes afaik

simo5 commented on pkg/auth/oauth/registry/registry_test.go:
```diff
@@ -473,3 +476,204 @@ func TestAuthenticateTokenValidated(t *testing.T) {
+	go timeouts.Run(stopCh)
+
+	// wait to see that the other thread has updated the timeouts
+	time.Sleep(1 * time.Second)
```
true, will probably move to use clock package in the future

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,215 @@
+	return nil
+}
+
+func (a *TimeoutValidator) updateToken(td *tokenData) {
```
not really a queue though ... perhaps slurpToken ? :-)

simo5 replied:
renamed to putToken

simo5 commented on pkg/auth/oauth/registry/timeoutvalidator.go:
```diff
@@ -0,0 +1,215 @@
+	return err
+}
+
+func (a *TimeoutValidator) flush(flushHorizon time.Time) {
```
ok added some

simo5 commented on pkg/cmd/server/api/types.go:
```diff
@@ -948,6 +948,9 @@ type TokenConfig struct {
 	AuthorizeTokenMaxAgeSeconds int32
 	// AccessTokenMaxAgeSeconds defines the maximum age of access tokens
 	AccessTokenMaxAgeSeconds int32
+	// AccessTokenInactivityTimeoutSeconds defines a default timeout for tokens.
```
done

simo5 commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -839,6 +839,9 @@ type TokenConfig struct {
 	AuthorizeTokenMaxAgeSeconds int32 `json:"authorizeTokenMaxAgeSeconds"`
 	// AccessTokenMaxAgeSeconds defines the maximum age of access tokens
 	AccessTokenMaxAgeSeconds int32 `json:"accessTokenMaxAgeSeconds"`
+	// AccessTokenInactivityTimeoutSeconds defines a default timeout for tokens.
```
done

simo5 commented on pkg/cmd/server/origin/authenticator.go:
```diff
@@ -92,6 +95,14 @@ func newAuthenticator(config configapi.MasterConfig, accessTokenGetter oauthclie
 	// OAuth token
 	if config.OAuthConfig != nil {
 		validators := []openshiftauthenticator.OAuthTokenValidator{authnregistry.NewExpirationValidator(), authnregistry.NewUIDValidator()}
+		if config.OAuthConfig.TokenConfig.AccessTokenInactivityTimeoutSeconds != nil {
```
ok

simo5 commented on pkg/oauth/apis/oauth/validation/validation.go:
```diff
@@ -16,7 +16,17 @@ import (
+	// This value is also currently used to define the maximum flush timeout.
+	// That is a flush will occur at least once every 5 minutes regardles of
+	// the actual default token's timeout interval.
+	ReferenceTimeoutInterval = 5 * 60
```
changed to MinimumInactivityTimeoutSeconds, updated description

simo5 commented on pkg/oauth/apis/oauth/validation/validation_test.go:
```diff
@@ -467,6 +498,22 @@ func TestValidateAccessTokensUpdate(t *testing.T) {
 			T: field.ErrorTypeInvalid,
 			F: "[]",
 		},
+		"change InactivityTimeoutSeconds to smaller value": {
```
done and fixed code to refer to inactivityTimeoutSeconds instead of timeoutsIn ... it be nice if those strings could be derived from code variable names via some macro system rather than have to type them in and get things wrong ...

enj commented:
Possible TODO: consider setting timeout == ExpiresIn if timeout > ExpiresIn (in token strategy)

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
Seem like we'll jave to wait for #17962 to merge before we can merge this.

simo5 commented:
/retest

simo5 commented:
/hold cancel

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
Flake #17781

simo5 commented:
/retest

simo5 commented:
@liggitt PTAL and LGTM if you are ok with this PR

simo5 commented:
flake: #17974

simo5 commented:
flake: #17731

simo5 commented:
flake:  #16484

simo5 commented:
/retest

enj commented:
/lgtm

Manually approving the apps changes since they are generated code.  David did not seem too concerned.

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
flake: #17985

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
flake: #16994

simo5 commented:
/retest

simo5 commented:
flake: #17697

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
/retest

simo5 commented:
flake: #15630
/retest

simo5 commented:
/retest
